subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/SelectorState.java

```diff
@@ -114,6 +115,10 @@ public ComponentState resolveModuleRevisionId() {
     public void restart(ComponentState moduleRevision) {
         this.selected = moduleRevision;
         this.targetModule = moduleRevision.getModule();
+        ComponentResolveMetadata metaData = moduleRevision.getMetaData();
+        if (metaData != null) {
+            this.idResolveResult.resolved(metaData);
```

Comment:
This change is also not required to make the test pass. Why is this required? Can you think of a test that would prevent us from breaking this again in the future?

Reply:
It is actually required. There's at least one test failing if we don't do this, because the serialized id is wrong if we don't do this.

Reply:
OK. I only ran the new test, and I was curious why this was required.